src/create-class.test.ts

```diff
@@ -0,0 +1,168 @@
+import { createClass } from "./create-class";
+import { Selection, Position } from "./editor";
+import {
+  createShouldUpdateCodeFor,
+  createShouldNotUpdateCodeFor
+} from "./test-helpers";
+
+const shouldUpdateCodeFor = createShouldUpdateCodeFor(createClass);
+const shouldNotUpdateCodeFor = createShouldNotUpdateCodeFor(createClass);
+
+describe("create class declaration from a new expression", () => {
+  it("with nothing else", () => {
+    shouldUpdateCodeFor({
+      code: "new ReadCode();",
+      selection: Selection.cursorAt(0, 0),
+      expected: {
+        code: `
+class ReadCode {
+  \${0:// Implement}
+}`,
+        position: new Position(1, 0),
+        name: 'Create class "ReadCode"'
+      }
+    });
+  });
+
+  it("assigned to a variable", () => {
+    shouldUpdateCodeFor({
+      code: "const code = new ReadCode();",
+      selection: Selection.cursorAt(0, 13),
+      expected: {
+        code: `
+class ReadCode {
+  \${0:// Implement}
+}`
+      }
+    });
+  });
+
+  it("assignment called with await", () => {
+    shouldUpdateCodeFor({
+      code:
+        "async function doSomethingAsync() { const code = await new ReadCode(); }",
+      selection: Selection.cursorAt(0, 60),
+      expected: {
+        code: `
+class ReadCode {
+  \${0:// Implement}
+}`
+      }
+    });
+  });
+
+  it("await without assignment", () => {
+    shouldUpdateCodeFor({
+      code: "async function doSomethingAsync() { await new ReadCode(); }",
+      selection: Selection.cursorAt(0, 45),
+      expected: {
+        code: `
+class ReadCode {
+  \${0:// Implement}
+}`
+      }
+    });
+  });
+
+  it("param of another call", () => {
+    shouldUpdateCodeFor({
+      code: "console.log(new ReadCode());",
+      selection: Selection.cursorAt(0, 13)
+    });
+  });
+
+  it("with assigned value referenced later", () => {
+    shouldUpdateCodeFor({
+      code: `const code = new ReadCode();
+write(code);`,
+      selection: Selection.cursorAt(0, 13),
+      expected: {
+        code: `
+class ReadCode {
+  \${0:// Implement}
+}
+
+`
```

Comment:
🍰 I'd also double-check the Position here because it matters

Reply:
as it stands the quick-fix will create the class as follows:
```
const code = new ReadCode();

class ReadCode {
  // Implement
}

write(code);
```
which of course doesn't make sense as `ReadCode` is used before being declared.

am I right in thinking that for a simple script like this we'd expect the class to be created above the code sample?

and for nested uses, we'd create the class at the root level with the position (before or after the code block) being less important?

Reply:
Indeed, the class should be declared before usage. That can be done by changing the snippet and the position where it's inserted.

For nested use it's a bit trickier since class should exist before the code is executed. But at this point I'd let the user deal with it and keep creating the class before the code block.